File: 0054-spiral-matrix/0054-spiral-matrix.java
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        int total = matrix.length * matrix[0].length;
        List<Integer> lst = new ArrayList<>();
        int col = matrix[0].length, row = matrix.length;
        int sc = 0, sr = 0;
        while (sc < col && sr < row) {
            for (int i = sc; i < col; i++) {
                lst.add(matrix[sr][i]);
            }
            sr++;
            for (int i = sr; i < row; i++) {
                lst.add(matrix[i][col-1]);
            }
            col--;
            if (sr < row) {
                for (int i = col-1; i >= sc; i--) {
                    lst.add(matrix[row-1][i]);
                }
                row--;
            }
            if (sc < col) {
                for (int i = row-1; i >= sr; i--) {
                    lst.add(matrix[i][sc]); 
                }
                sc++;
            }
        }
        return lst;
    }
}

File: 0054-spiral-matrix/0054-spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        row = len(matrix)
        column = len(matrix[0])
        length = row * column
        row_1 = 0
        column_1 = 0
        i = 0
        j = 0
        ans = []
        print(row * column)
        while len(ans) < (length):
            print(len(ans))
            while j < column and len(ans) < (length):
                print(i, j)
                ans.append(matrix[i][j])
                j += 1
            column -= 1
            j -= 1
            i += 1
            while i < row and len(ans) < (length):
                print(i, j)
                ans.append(matrix[i][j])
                i += 1
            row -= 1
            i -= 1
            j -= 1
            while j >= column_1 and len(ans) < (length):
                print(i, j)
                ans.append(matrix[i][j])
                j -= 1
            column_1 += 1
            j += 1
            i -= 1
            while i > row_1 and len(ans) < (length):
                print(i, j)
                ans.append(matrix[i][j])
                i -= 1
            row_1 += 1
            i += 1
            j += 1
            print(len(ans))
        return ans
